hwbench/environment/vendors/monitoring_device.py

```diff
@@ -0,0 +1,90 @@
+import cachetools.func
+import json
+import logging
+import redfish  # type: ignore
+from ...utils import helpers as h
+from ...bench.monitoring_structs import (
+    MonitorMetric,
+)
+from typing import Any
+
+
+class MonitoringDevice:
+    def __init__(self, vendor):
+        self.vendor = vendor
+        self.redfish_obj = None
+        self.logged = False
+
+    def __del__(self):
+        if self.logged:
+            self.redfish_obj.logout()
+
+    def add_monitoring_value(
+        self,
+        monitoring_struct: dict[str, dict[str, MonitorMetric]],
+        context: Any,
+        metric: MonitorMetric,
+        name: str,
+        value: float,
+    ) -> dict[str, dict[str, MonitorMetric]]:
+        """This function add a new <value> in the monitoring data structure."""
+        if str(context) not in monitoring_struct:
+            monitoring_struct[str(context)] = {}
+        if name not in monitoring_struct[str(context)]:
+            monitoring_struct[str(context)][name] = metric
+        monitoring_struct[str(context)][name].add(value)
+        return monitoring_struct
+
+    def get_driver_name(self) -> str:
+        """Return the driver name"""
+        return type(self).__name__
+
+    def dump(self) -> dict[str, str]:
+        """Return the dump of the drive"""
+        return {"driver": self.get_driver_name()}
+
+    def connect_redfish(self, username: str, password: str, device_url: str):
+        """Connect to the device using Redfish."""
+
+        try:
+            if "https://" not in device_url:
+                device_url = "https://{}".format(device_url)
+            self.redfish_obj = redfish.redfish_client(
+                base_url=device_url,
+                username=username,
+                password=password,
+                default_prefix="/redfish/v1",
+                timeout=10,
+            )
+            self.redfish_obj.login()
+            self.logged = True
+        except json.decoder.JSONDecodeError:
+            h.fatal("JSONDecodeError on {}".format(device_url))
+        except redfish.rest.v1.RetriesExhaustedError:
+            h.fatal("RetriesExhaustedError on {}".format(device_url))
+        except redfish.rest.v1.BadRequestError:
+            h.fatal("BadRequestError on {}".format(device_url))
+        except redfish.rest.v1.InvalidCredentialsError:
+            h.fatal("Invalid credentials for {}".format(device_url))
+        except Exception as exception:
+            h.fatal(type(exception))
+
+    @cachetools.func.ttl_cache(maxsize=128, ttl=1.5)
```

Comment:
Do you think we will want the same caching properties for both PDUs and BMCs?

Reply:
I'd think so, the concept is that we don't really need high precision since the device already manages an aggregation. This cache ensures that we don't overpoll devices so I think that's valid for all devices, that's why I put it here .